Gsoc Assignment vertx-http-proxy pr2(ramanadas)

ramanadas opened the pull request:
made changes to commits

ramanadas committed:
Fetching jokes

backend on port 8080

ramanadas committed:
Proxy

proxy server port is 7070

ramanadas committed:
User interface

frontend server on port 3000

ramanadas commented:
hi @tsegismont made changes in commits. please check this out.

tsegismont commented on gsoc-2024-backend/src/main/java/backend/BackendVerticle.java:
```diff
@@ -11,10 +17,46 @@ public static void main(String[] args) {
+  }
+  
+  public void getJoke(Promise<String > promise) {
+    WebClient client = WebClient.create(vertx);
```
This will create a client (and a pool of connections) for each request. Can you provide a better solution?

ramanadas replied:
i will move it to start().thanks for info.

tsegismont commented on gsoc-2024-backend/src/main/java/backend/BackendVerticle.java:
```diff
@@ -11,10 +17,46 @@ public static void main(String[] args) {
+
+  }
+  
+  public void getJoke(Promise<String > promise) {
```
Since `client.send()` returns a `Future`, can your adapt the implementation to transform the future instead of using an intermediary promise?

ramanadas replied:
i will try tp remove intermediary promise.

tsegismont commented on gsoc-2024-backend/src/main/java/backend/BackendVerticle.java:
```diff
@@ -11,10 +17,46 @@ public static void main(String[] args) {
+      });
+    });
+
+    server.requestHandler(router).listen(8080);
```
What happens if the server does not start correctly? Can you update this to complete/fail the `startPromise` as needed?

ramanadas replied:
i will handle this case.

tsegismont commented on gsoc-2024-backend/src/main/java/backend/BackendVerticleTest.java:
```diff
@@ -0,0 +1,63 @@
+    // Deploy the BackendVerticle
+    vertx.deployVerticle(new BackendVerticle(), testContext.succeeding(id -> testContext.completeNow()));
+
+    // Create an HTTP server and client for testing
+    server = vertx.createHttpServer();
+    client = vertx.createHttpClient();
+    server.requestHandler(req -> {
+      if (req.method() == HttpMethod.GET && req.path().equals("/joke")) {
+        req.response().end("Test joke");
+      } else {
+        req.response().setStatusCode(404).end();
+      }
+    }).listen(8080, testContext.succeeding(result -> testContext.completeNow()));
```
Whats is this server needed for?

ramanadas replied:
its actuvally not needed. i forget to remove.

tsegismont commented on gsoc-2024-proxy/src/main/java/proxy/ProxyVerticle.java:
```diff
@@ -15,6 +22,24 @@ public static void main(String[] args) {
+
+    HttpServer proxyServer = vertx.createHttpServer();
+    Router proxyRouter = Router.router(vertx);
+    proxyRouter.route()
+        .handler(
+          CorsHandler.create()
+            .addRelativeOrigin("http://localhost:3000")
+            .allowedMethod(HttpMethod.GET)
+        );
```
Why is Cors needed? Your Proxy should send the requests for `/joke` to the backend service, and the others to the website service.

ramanadas replied:
sorry for that. i misunderstand that proxy is for only backed server. so cors needed for fronted. now i will modify for frondend proxy.

tsegismont reviewed (commented):
Thank you. Here is a first batch of comments

ramanadas committed:
fixed issues

client object now created only once.
added failed server response.
and removed promise and unwanted lines in tesr cases.

ramanadas committed:
Update ProxyVerticle.java

updated proxy for frontend also.

ramanadas committed:
Update WebsiteVerticle.java

added failed server response

ramanadas commented:
Updated can you check it @tsegismont

ramanadas commented:
Can you provide me information related to proposal format and what kind of information needed in that document

tsegismont commented on gsoc-2024-backend/src/main/java/backend/BackendVerticle.java:
```diff
@@ -1,20 +1,72 @@
+        startPromise.fail(result.cause());
+      }
+    });
+    startPromise.future().onSuccess(h->{
+      System.out.println("server deployed successfully");
+    })
+      .onFailure(h->{
+        System.out.println("failed to deploy server");
+      });
```
In fact, you should complete the `startPromise` (or fail it, in case of a problem)

ramanadas replied:
,in 62 line startPromise.fail

tsegismont replied:
Ok, I missed this.

tsegismont commented on gsoc-2024-backend/src/main/java/backend/BackendVerticle.java:
```diff
@@ -1,20 +1,72 @@
+  }
+
+  public Future<String> getJoke() {
+    Promise<String> promise = Promise.promise();
```
Can you try to get rid of this completely?

ramanadas replied:
i removed it. used future compose instead.

ramanadas commented on gsoc-2024-backend/src/main/java/backend/BackendVerticle.java:
```diff
@@ -1,20 +1,72 @@
+      if(result.succeeded()){
+        startPromise.complete();
+      } else {
+        startPromise.fail(result.cause());
```
Here I fail startPromise

ramanadas committed:
removed promise

used future compose alternatively

ramanadas commented:
> Can you provide me information related to proposal format and what kind of information needed in that document

@tsegismont

tsegismont reviewed (approved):
Looks good to me, thanks

tsegismont commented:
> > Can you provide me information related to proposal format and what kind of information needed in that document
> 
> @tsegismont

Google provides a guide: https://google.github.io/gsocguides/student/writing-a-proposal

tsegismont commented:
Please send me your proposal by email instead of posting it here. Thanks

ramanadas commented:
Thanks you

ramanadas committed:
Updated for readability

improved readability
